File: backend/javascript/src/services/search/search.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { EmbeddingBuilder } from '../ingestion/processors/embedding.builder';
import { VectorStoreManager } from '../providers/vectorstore/vector-store-manager';
import { SearchResult } from '../providers/vectorstore/vector-store';

/**
 * [검색 서비스]
 * 
 * 사용자의 질문을 임베딩하고 벡터 저장소에서 유사한 조각을 찾아주는 
 * 비즈니스 로직을 담당합니다.
 */
@Injectable()
export class SearchService {
    private readonly logger = new Logger(SearchService.name);

    constructor(
        private readonly embeddingBuilder: EmbeddingBuilder,
        private readonly vectorStore: VectorStoreManager,
    ) { }

    /**
     * 사용자의 질문을 받아 가장 유사한 문서 조각들을 찾아 반환합니다.
     */
    async execute(query: string, topK: number = 3): Promise<SearchResult[]> {
        this.logger.log(`Searching for: "${query}"`);

        // 1. 질문(문장)을 임베딩(숫자)으로 변환합니다.
        // 우리 시스템의 embed 메소드는 배열을 받으므로 임시 객체를 만들어 전달합니다.
        const dummyChunk: any = { text: query, metadata: {} };
        const [embeddedQuery] = await this.embeddingBuilder.embed([dummyChunk]);

        const vector = embeddedQuery.metadata.embedding;
        if (!vector) {
            this.logger.error("[Search] Query embedding failed! Vector is null.");
            return [];
        }

        // 2. 변환된 숫자 좌표를 가지고 벡터 저장소에서 검색합니다.
        const results = await this.vectorStore.search(
            vector,
            topK
        );

        this.logger.log(`[Search] Found ${results.length} results. Top score: ${results[0]?.score ?? 'N/A'}.`);
        return results;
    }
}

File: backend/java/src/main/java/com/ragassistant/service/SearchService.java
package com.ragassistant.service;

import com.ragassistant.model.DocumentChunk;
import com.ragassistant.provider.embedding.EmbeddingBuilder;
import com.ragassistant.provider.vectorstore.ChromaVectorStore;
import com.ragassistant.provider.vectorstore.VectorStore;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.Collections;
import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
public class SearchService {
    private final EmbeddingBuilder embeddingBuilder;
    private final ChromaVectorStore vectorStore;

    public List<VectorStore.SearchResult> execute(String query, int topK) {
        log.info("Searching for: \"{}\"", query);

        // 1. Embedding the query
        DocumentChunk dummy = DocumentChunk.builder()
                .text(query)
                .metadata(DocumentChunk.ChunkMetadata.builder().build())
                .build();

        List<DocumentChunk> embedded = embeddingBuilder.embed(Collections.singletonList(dummy));
        List<Double> vector = embedded.get(0).getMetadata().getEmbedding();

        if (vector == null || vector.isEmpty()) {
            log.error("[Search] Query embedding failed!");
            return Collections.emptyList();
        }

        // 2. Search in Vector Store
        List<VectorStore.SearchResult> results = vectorStore.search(vector, topK);
        log.info("[Search] Found {} results.", results.size());

        for (int i = 0; i < results.size(); i++) {
            log.info(results.get(i).getChunk().getText());
        }

        return results;
    }
}
